keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,172 @@
+
+In this proposal, we choose to _only_ trace phases of an object's lifecycle wherein it's correcting from an undesired state to its desired state, and to end the trace when it enters this desired state. This means that the same object will export traces for each reconciliation it undergoes. This decision was made because:
+
+1) These reconciliation phases are where Kubernetes performs actual work that can be traced
+2) Prevents traces from spanning the entire life of an object, which can be indefninite
+3) Provides a consistent, concrete duration for the root span which remains applicable across object types
+4) The time required for Kubernetes controllers to reconcile desired and actual state is, by definition, a measurement of either latency experienced by end-users in the fulfillment of their requests, or down-time due to a disruption
+
+#### Root spans
+
+In the standard model for distributed tracing, there exists a span in each trace that all other spans are descendents of and which extends the length of the entire trace, called the `root span`. 
+
+The Kubernetes component that kicks off an operation might not be the same component that ends it. In this proposal, when we are at the point where we want to end a root span, we craft a span to export which acts as the root span for the trace. For example, when the kubelet updates a pod from `Pending` to `Running`, it creates a root span using the start time of the pod as the start, and the current time as the end.
+
+
+#### Context propagation
+
+To correlate work done between components as belonging to the same trace, we must pass span context across process boundaries. In traditional distributed systems, this context can be passed down through RPC metadata or HTTP headers. Kubernetes, however, due to its watch-based nature, requires us to attach trace context directly to the target object. 
+
+In this proposal, we choose to propagate this span context as an encoded string an object annotation called `trace.kubernetes.io/context`. This annotation value is regenerated and replaced when an object's trace ends, to achieve the desired behavior from [section one](#trace-lifecycle). 
+
+This proposal chooses to use annotations as a less invasive alternative to adding a field to object metadata, but as this proposal matures, adding trace context to the official API should be considered. 
+
+
+### In-tree changes
+
+#### Vendoring the OpenCensus trace framework
+
+This KEP proposes the use of the [OpenCensus tracing framework](https://opencensus.io/) to create and export spans to configured backends. The OpenCensus framework was chosen for various reasons: 
+
+1) Provides concrete, tested implementations for creating and exporting spans to diverse backends, rather than providing an API specification, as is the case with [OpenTracing](https://opentracing.io/specification/)
```

Comment:
I'm generally a big fan of the motivations and intentions of the OpenCensus project, but I'm a little concerned about it being a rather young project.

Reply:
Agreed, the OC project is still quite young. However, based on the fact that this would be an experimental, opt-in alpha feature, it might be acceptable for us to bring in for use provided we stick to its stable features (starting, ending, and exporting spans).